src/ledger/LedgerTxnExpirationSQL.cpp

```diff
@@ -0,0 +1,382 @@
+        }
+    }
+#endif
+};
+
+void
+LedgerTxnRoot::Impl::bulkDeleteExpiration(
+    std::vector<EntryIterator> const& entries, LedgerTxnConsistency cons)
+{
+    BulkDeleteExpirationOperation op(mApp.getDatabase(), cons, entries);
+    mApp.getDatabase().doDatabaseTypeSpecificOperation(op);
+}
+
+class BulkUpsertExpirationOperation : public DatabaseTypeSpecificOperation<void>
+{
+    Database& mDb;
+    std::vector<std::string> mKeyHashes;
+    std::vector<std::string> mExpirationEntries;
+    std::vector<int32_t> mLastModifieds;
+
+    void
+    accumulateEntry(LedgerEntry const& entry)
+    {
+        throwIfNotExpiration(entry.data.type());
+
+        mKeyHashes.emplace_back(
+            toOpaqueBase64(entry.data.expiration().keyHash));
+        mExpirationEntries.emplace_back(toOpaqueBase64(entry));
+        mLastModifieds.emplace_back(
+            unsignedToSigned(entry.lastModifiedLedgerSeq));
+    }
+
+  public:
+    BulkUpsertExpirationOperation(Database& Db,
+                                  std::vector<EntryIterator> const& entryIter)
+        : mDb(Db)
+    {
+        for (auto const& e : entryIter)
+        {
+            releaseAssert(e.entryExists());
+            accumulateEntry(e.entry().ledgerEntry());
+        }
+    }
+
+    void
+    doSociGenericOperation()
+    {
+        std::string sql = "INSERT INTO expiration "
+                          "(keyhash, ledgerentry, lastmodified) "
+                          "VALUES "
+                          "( :keyHash, :v1, :v2 ) "
+                          "ON CONFLICT (keyhash) DO UPDATE SET "
+                          "ledgerentry = excluded.ledgerentry, "
+                          "lastmodified = excluded.lastmodified";
+
+        auto prep = mDb.getPreparedStatement(sql);
+        soci::statement& st = prep.statement();
+        st.exchange(soci::use(mKeyHashes));
+        st.exchange(soci::use(mExpirationEntries));
+        st.exchange(soci::use(mLastModifieds));
+        st.define_and_bind();
+        {
+            auto timer = mDb.getUpsertTimer("expiration");
+            st.execute(true);
+        }
+        if (static_cast<size_t>(st.get_affected_rows()) != mKeyHashes.size())
+        {
+            throw std::runtime_error("Could not update data in SQL");
+        }
+    }
+
+    void
+    doSqliteSpecificOperation(soci::sqlite3_session_backend* sq) override
+    {
+        doSociGenericOperation();
+    }
+
+#ifdef USE_POSTGRES
+    void
+    doPostgresSpecificOperation(soci::postgresql_session_backend* pg) override
+    {
+        std::string strKeyHashes, strExpirationEntries, strLastModifieds;
+
+        PGconn* conn = pg->conn_;
+        marshalToPGArray(conn, strKeyHashes, mKeyHashes);
+        marshalToPGArray(conn, strExpirationEntries, mExpirationEntries);
+        marshalToPGArray(conn, strLastModifieds, mLastModifieds);
+
+        std::string sql = "WITH r AS "
+                          "(SELECT unnest(:v1::TEXT[]), "
```

Comment:
for legibility (not correctness) you might want to change this so :v1 doesn't occur twice in the query. SOCI ignores it IIRC but y'know, it's a little confusing to a reader.